README.md
# tindetheus
Build personalized machine learning models for Tinder based on your historical preference using Python.

There are three parts to this:
1. A function to build a database which records everything about the profiles you've liked and disliked.
2. A function to train a model to your database.
3. A function to use the trained model to automatically like and dislike new profiles.

# How it works
The last layer of a CNN trained for facial classification can be used as a feature set which describes an individual's face. It just so happens that this feature set is related to facial attractiveness.

tindetheus let's you build a database based on the profiles that you like and dislike. You can then train a classification model to your database. The model training first uses a MTCNN to detect and box the faces in your database. Then a facenet model is run on the faces to extract the embeddings (last layer of the CNN). A logistic regression model is then fit to the embeddings. The logistic regression model is saved, and this processes is repeated in automation to automatically like and dislike profiles based on your historical preference.

![Visual aid explaining tindetheus](https://raw.githubusercontent.com/cjekel/tindetheus/master/examples/how_does_tindetheus_work.png)

This [blog post](http://jekel.me/2018/Using-facenet-to-automatically-like-new-tinder-profiles/) has a short description of how tindetheus works.

For a more detailed description of how and why this works see [https://arxiv.org/abs/1803.04347](https://arxiv.org/abs/1803.04347)

# Example usage
```bash
tindetheus browse
```
build a database by liking and disliking profiles on Tinder. The database contains all the profile information as a numpy array, while the profile images are saved in a different folder.

```bash
tindetheus browse --distance=20
```
by default tindetheus starts with a 5 mile radius, but you can specify a search distance by specifying --distance. The above example is to start with a 20 mile search radius. It is important to note that when you run out of nearby users, tindethesus will ask you if you'd like to increase the search distance by 5 miles.

```bash
tindetheus train
```
Use machine learning to build a personalized model of who you like and dislike based on your database. The more profiles you've browsed, the better your model will be.

```bash
tindetheus like
```
Use your personalized model to automatically like and dislike profiles. The profiles which you have automatically liked and disliked are stored in al_database. By default this will start with a 5 mile search radius, which increases by 5 miles until you've used 100 likes. You can change the default search radius by using
```bash
tindetheus like --distance=20
```
which would start with a 20 mile search radius.

# Installation and Getting started
Installation and getting started guide now stored in [GETTING_STARTED.md](https://github.com/cjekel/tindetheus/blob/master/GETTING_STARTED.md)

# .env
You can now store all default optional parameters in your environment variables! This means you can set your starting distance, number of likes, and image_batch size without manually specifying the options each time. This is an example `.env` file:
```
FACEBOOK_AUTH_TOKEN="TODO" # your facebook token hash
# alternatively you can use the XAuthToken
# TINDER_AUTH_TOKEN="TODO"
TINDETHEUS_MODEL_DIR="/models/20170512-110547"  # the location of your facenet model directory
# see https://github.com/davidsandberg/facenet#pre-trained-models for other
# pretrained facenet models
TINDETHEUS_IMAGE_BATCH=1000  # number of images to load in a batch during train
#  the larger the image_batch size, the faster the training process, at the
#  cost of additional memory. A 4GB machine may struggle with 1000 images.
TINDETHEUS_DISTANCE=5  # Set the starting distance in miles
TINDETHEUS_LIKES=100  # set the number of likes you want to use
#  note that free Tinder users only get 100 likes in 24 hours
TINDETHEUS_RETRIES=20
```

# Using the validate function on a different dataset
As of Version 0.4.0, tindetheus now includes a validate function. This validate functions applies your personally trained tinder model on an external set of images. If there is a face in the image, the model will predict whether you will like or dislike this face. The results are saved in validation.csv. For more information about the validate function [read this](https://github.com/cjekel/tindetheus/blob/master/VALIDATE_GUIDE.md).

# Dataset available upon request

The dataset used to create this work is available upon request. Please fill out this [form](https://forms.gle/tTVZVxcvtwRm1w659) to request access to the data.

# Changelog
All changes now stored in [CHANGELOG.md](https://github.com/cjekel/tindetheus/blob/master/CHANGELOG.md)

# Open source libraries
tindetheus uses the following open source libraries:

- [pynder](https://github.com/charliewolf/pynder)
- [facenet](https://github.com/davidsandberg/facenet)
- [numpy](http://www.numpy.org/)
- [matplotlib](https://matplotlib.org/)
- [scikit-learn](http://scikit-learn.org/stable/)
- [tensorflow](https://www.tensorflow.org/)
- [imageio](https://imageio.github.io/)
- [pandas](http://pandas.pydata.org/)

# About the name
Tindetheus is a combination of Tinder (the popular online dating application) and the Greek Titans: [Prometheus](https://en.wikipedia.org/wiki/Prometheus) and [Epimetheus](https://en.wikipedia.org/wiki/Epimetheus_(mythology)). Prometheus signifies "forethought," while  his brother Epimetheus denotes "afterthought". In synergy they serve to improve your Tinder experience.

Epimetheus creates a database from all of the profiles you review on Tinder.

Prometheus learns from your historical preferences to automatically like new Tinder profiles.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.2] - 2020-12-09
### Added
- Note in readme that the dataset used to create this work is available upon request by filling out this [form](https://forms.gle/tTVZVxcvtwRm1w659) 

## [1.0.1] - 2020-05-24
### Changed
- Bugfix: config.py was setting the retries env variable as the like env variable; this has been corrected
- Bugfix: remove string endings from .env variable strings if they exist

## [1.0.0] - 2020-03-28
### Added
- You can now store all default optional parameters in your environment variables! This means you can set your starting distance, number of likes, and image_batch size without manually specifying the options each time. You can locally set a `.env` file within your tinder working directory. Check out this [.env](https://github.com/cjekel/tindetheus/blob/master/.env.example) example file. Thanks to Vikash Kothary for this contribution.
- .env file will work on Windows/Linux/Mac!
### Removed
- config.txt file no longer supported! Switch to a .env file, or set your environment variable!
### Changed
- New single sourced version file is within setup.py.

## [0.5.5] - 2019-12-04
### Changed
- Fix Python 2.7 related issues with login and config.txt parsing

## [0.5.4] - 2019-11-27
### Added
- retries option for tindetheus like or browse command, default to 20, and can be modified in config.txt, for example ```retries = 25```
### Changed
- cleaned up some code for parsing config.txt, seperated interger and string types
- setup.py now lists ```tensorflow < 2.0.0``` as a requirement

## [0.5.3] - 2019-11-17
### Changed
- skimage.transform.resize now uses defaults

## [0.5.2] - 2019-11-13
### Added
- setuptools now shows up in the requirements

## [0.5.1] - 2019-11-07
### Changed
- fixed a bug that would cause tindetheus not to run if a line in config.txt was a list of 2 items when separated by a space
- switched to setuptools, remove README.rst

## [0.5.0] - 2019-07-28
### Added
- Basic travis.ci checking for installation on python 2.7, 3.5, 3.6 and flake8 on files
- Reorganize tindetheus functions into submodules image_processing, machine_learning, tinder_client
### Changed
- Bugfix python 2.7 installation
- Bugfix related to try except that was overriding the database folder on newer numpy distributions
- Moved installation instructions, changelog, and validate function details to separate files
- numpy.load now uses allow_pickle=True
### Removed
- Remove function from tindetheus into submodules image_processing, machine_learning, tinder_client

## [0.4.7] - 2019-07-27
### Changed
- Fix installation on Windows
- Fix loading MTCNN weights on newer version of numpy

## [0.4.6] - 2019-06-23
### Added
- Docker container instructions.
### Changed
- Readme notes 
- Bugfix related to Python 2.7 command line parsing

## [0.4.3] - 2019-05-05
### Added
- Add option to log in using XAuthToken thanks to charlesduponpon
- Add like_folder command line option to create al/like and al/dislike folders based on the historically liked and disliked profiles (this is a quick way to assess model quality)

## [0.4.1] - 2019-04-29
### Changed
- Fix issue where line endings that were causing authentication failure
- Fix handling of config.txt

## [0.4.0] - 2018-12-02
### Added
- New validate function to apply your tindetheus model to a new dataset. See README on how to use this function.
### Changed
- Fix issues with lossy integer conversions
- Some other small bug fixes

## [0.3.3] - 2018-11-25
### Changed
- Update how facenet TensorFlow model is based into object
- Fixes session recursion limit

## [0.3.2] - 2018-11-04
### Changed
- tindetheus will now exit gracefully if you have used all of your free likes while running tindetheus like

## [0.3.1] - 2018-11-04
### Changed
- Fix bug related to Windows and calc_avg_emb(), which wouldn't find the unique classes

## [0.3.0] - 2018-11-03
### Added
- Added version tracking and parser with --version
- New optional parameters: likes (set how many likes you have remaining default=100), and image_batch (set the number of images to load into facenet when training default=1000)
- Now all optional settings can be saved in config.txt
### Changes
- Bug fix related to calling a tindetheus.export_embeddings function
- Saving the same filename in your database no longer bombs out on Windows
- Code should now follow pep8

## [0.2.11] - 2018-05-11
### Added
- Added support for latest facenet models. The different facenet models don't appear to really impact the accuracy according to [this post](https://jekel.me/2018/512_vs_128_facenet_embedding_application_in_Tinder_data/).
- You can now specify which facenet model to use in the config.txt file.
- Added [example](https://github.com/cjekel/tindetheus/blob/master/examples/open_database.py) script for inspecting your database manually
### Changed
- Updated facenet clone implementation
- Now requires minimum tensorflow version of 1.7.0


tindetheus/tindetheus.py
# MIT License
#
# Copyright (c) 2017-2019 Charles Jekel
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import sys
import os
import argparse

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
import joblib

try:
    import urllib
    urllib_HTTP_Error = urllib.error.HTTPError
except AttributeError:
    import urllib2
    urllib_HTTP_Error = urllib2.HTTPError

from .version import __version__  # noqa F401
from tindetheus import config
from tindetheus import export_embeddings
from tindetheus import tindetheus_align
from tindetheus.tinder_client import client
import tindetheus.facenet_clone.facenet as facenet
import tindetheus.image_processing as imgproc
import tindetheus.machine_learning as ml

try:
    FileNotFoundError
except NameError:
    FileNotFoundError = IOError


def catch_http_error(e, args, retries, facebook_token, x_auth_token):
    print(e)
    print("Closing session.")
    retries -= 1
    if retries > 0:
        print("Let's try again. \n Retries left:", retries)
        main(args, facebook_token, x_auth_token=x_auth_token,
             retries=retries)
    else:
        print("Out of retries. Exiting.")


def main(args, facebook_token, x_auth_token=None, retries=20):
    # There are three function choices: browse, build, like
    # browse: review new tinder profiles and store them in your database
    # train: use machine learning to create a new model that likes and dislikes
    # profiles based on your historical preference
    # like: use your machine leanring model to like new tinder profiles
    if args.function == 'browse':
        try:
            my_sess = client(facebook_token, args.distance, args.model_dir,
                             likes_left=args.likes, x_auth_token=x_auth_token)
            my_sess.browse()
        except urllib_HTTP_Error as e:
            catch_http_error(e, args, retries, facebook_token, x_auth_token)
    elif args.function == 'train':
        # align the database
        tindetheus_align.main()
        # export the embeddings from the aligned database
        export_embeddings.main(model_dir=args.model_dir,
                               image_batch=args.image_batch)
        # calculate the n average embedding per profiles
        X, y = ml.calc_avg_emb()
        # fit and save a logistic regression model to the database
        ml.fit_log_reg(X, y)

    elif args.function == 'validate':
        print('\n\nAttempting to validate the dataset...\n\n')
        valdir = 'validation'
        # align the validation dataset
        tindetheus_align.main(input_dir=valdir,
                              output_dir=valdir+'_aligned')
        # export embeddings
        # y is the image list, X is the embedding_array
        image_list, emb_array = export_embeddings.main(model_dir=args.model_dir,  # noqa E501
                                                       data_dir=valdir+'_aligned',  # noqa E501
                                                       image_batch=args.image_batch,  # noqa E501
                                                       embeddings_name='val_embeddings.npy',  # noqa E501
                                                       labels_name='val_labels.npy',  # noqa E501
                                                       labels_strings_name='val_label_strings.npy',  # noqa E501
                                                       return_image_list=True)  # noqa E501
        # print(image_list)
        # convert the image list to a numpy array to take advantage of
        # numpy array slicing
        image_list = np.array(image_list)
        print('\n\nEvaluating trained model\n \n')
        model = joblib.load('log_reg_model.pkl')
        yhat = model.predict(emb_array)
        # print(yhat)
        # 0 should be dislike, and 1 should be like
        # if this is backwards, there is probably a bug...
        dislikes = yhat == 0
        likes = yhat == 1
        imgproc.show_images(image_list[dislikes], holdon=True, title='Dislike')
        print('\n\nGenerating plots...\n\n')
        plt.title('Dislike')

        imgproc.show_images(image_list[likes], holdon=True, title='Like')
        plt.title('Like')

        cols = ['Image name', 'Model prediction (0=Dislike, 1=Like)']
        results = np.array((image_list, yhat)).T
        print('\n\nSaving results to validation.csv\n\n')
        my_results_DF = pd.DataFrame(results, columns=cols)
        my_results_DF.to_csv('validation.csv')

        plt.show()

    elif args.function == 'like':
        try:
            print('... Loading the facenet model ...')
            print('... be patient this may take some time ...')
            with tf.Graph().as_default():
                with tf.Session() as sess:
                    # pass the tf session into client object
                    my_sess = client(facebook_token, args.distance,
                                     args.model_dir, likes_left=args.likes,
                                     tfsess=sess, x_auth_token=x_auth_token)
                    # Load the facenet model
                    facenet.load_model(my_sess.model_dir)
                    print('Facenet model loaded successfully!!!')
                    # automatically like users
                    my_sess.like()
        except urllib_HTTP_Error as e:
            catch_http_error(e, args, retries, facebook_token, x_auth_token)

    elif args.function == 'like_folder':
        print('Copying al_database profiles into either al/like or al/dislike')
        # make folders
        if not os.path.exists('al'):
            os.makedirs('al')
        if not os.path.exists('al/like'):
            os.makedirs('al/like')
        if not os.path.exists('al/dislike'):
            os.makedirs('al/dislike')

        # load the auto like database
        al_data = np.load('al_database.npy', allow_pickle=True)

        # copy profile images to either al/like or al/dislike
        for user in al_data:
            imgproc.al_copy_images(user[8], user[0], user[-1])

    else:
        text = '''You must specify a function. Your choices are either
tindetheus browse
tindetheus train
tindetheus like
tindetheus validate'''
        print(text)


def parse_arguments(argv, defaults):
    help_text = '''There are four function choices: browse, train, like, or validate.
\n

1) tindetheus browse
-- Let's you browse tinder profiles to add to your database.
-- Browses tinder profiles in your distance until you run out.
-- Asks if you'd like to increase the distance by 5 miles.
-- Use to build a database of the tinder profiles you look at.
\n
2) tindetheus train
-- Trains a model to your Tinder database.
-- Uses facenet implementation for facial detection and classification.
-- Saves logistic regression model to classify which faces you like and
-- dislike.
\n
3) tindetheus like
-- Automatically like and dislike Tinder profiles based on your historical
-- preference. First run browse, then run train, then prosper with like.
-- Uses the trained model to automatically like and dislike profiles.
-- Profiles where a face isn't detected are automatically disliked.
\n
4) tindetheus validate
-- This validate functions applies your personally trained tinder model on
-- an external set of images. Place images you'd like to run tindetheus on
-- withing a folder within the validation directory. See README for more
-- details. The results are saved in validation.csv.
\n
5) tindetheus like_folder
-- Creates al/like and al/dislike folders based on the profiles you have
-- automatically liked. This copies the profile images from al_database
-- into al/like or al/disliked based on whether the model liked or
-- disliked the profile.
\n
You can now store all default optional parameters in your environment
variables! This means you can set your starting distance, number of likes, and
image_batch size without manually specifying the options each time. You can
set local environment variables for a local folder using a .env file. This is
an example .env file:
FACEBOOK_AUTH_TOKEN="TODO" # your facebook token hash
# alternatively you can use the XAuthToken
# TINDER_AUTH_TOKEN="TODO"
TINDETHEUS_MODEL_DIR="/models/20170512-110547"
# the location of your facenet model directory
# see https://github.com/davidsandberg/facenet#pre-trained-models for other
# pretrained facenet models
TINDETHEUS_IMAGE_BATCH=1000  # number of images to load in a batch during train
#  the larger the image_batch size, the faster the training process, at the
#  cost of additional memory. A 4GB machine may struggle with 1000 images.
TINDETHEUS_DISTANCE=5  # Set the starting distance in miles
TINDETHEUS_LIKES=100  # set the number of likes you want to use
#  note that free Tinder users only get 100 likes in 24 hours
TINDETHEUS_RETRIES=20
\n
Optional arguments will overide config.txt settings.
'''
    parser = argparse.ArgumentParser(description=help_text,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)  # noqa: E501
    parser.add_argument('function', type=str,
                        help='browse, train, like, validate, or like_folder')
    parser.add_argument('--distance', type=int,
                        help='Set the starting distance in miles.'
                        'Tindetheus will crawl in 5 mile increments from here'
                        '. Default=5.', default=defaults['distance'])
    parser.add_argument('--image_batch', type=int,
                        help='The number of images to load in the facenet'
                        ' model at one time. This only affects the train '
                        'functionality. A larger number will be faster at the'
                        ' cost for larger memory. Default=1000.',
                        default=defaults['image_batch'])
    parser.add_argument('--model_dir', type=str, help='Location of your '
                        'pretrained facenet model. Default="20170512-110547"',
                        default=defaults['model_dir'])
    parser.add_argument('--likes', type=int, help='Set the number of likes to '
                        'use. Note that free Tinder users only get 100 likes '
                        'in 24 hour period', default=defaults['likes'])
    parser.add_argument('--version', action='version', version=__version__)
    return parser.parse_args(argv)


def command_line_run():
    # settings to look for
    defaults = {'facebook_token': config.FACEBOOK_AUTH_TOKEN,
                'XAuthToken': config.TINDER_AUTH_TOKEN,
                'model_dir': config.TINDETHEUS_MODEL_DIR,
                'image_batch': config.TINDETHEUS_IMAGE_BATCH,
                'distance': config.TINDETHEUS_DISTANCE,
                'likes': config.TINDETHEUS_LIKES,
                'retries': config.TINDETHEUS_RETRIES}
    # parse the supplied arguments
    args = parse_arguments(sys.argv[1:], defaults)

    if defaults['facebook_token'] is None and defaults['XAuthToken'] is None:
        raise('ERROR: No facebook or tinder token has been set.'
              'You must supply an auth token in order to use tindetheus!')

    # run the main function with parsed arguments
    main(args, defaults['facebook_token'], defaults['XAuthToken'],
         retries=defaults['retries'])


if __name__ == '__main__':
    command_line_run()
